canonical/cloud-init
# Initial hotplug support (SC-252)

## Proposed Commit Message


```
Initial hotplug support

Adds a udev script which will invoke a hotplug hook script on all net
add events. The script will write some udev arguments to a systemd FIFO
socket (to ensure we have only instance of cloud-init running at a
time), which is then read by a new service that calls a new 'cloud-init
devel hotplug-hook' command to handle the new event.

This hotplug-hook command will:
- Fetch the pickled datsource
- Verify that the hotplug event is supported/enabled
- Update the metadata for the datasource
- Ensure the hotplugged device exists within the datasource
- Apply the config change on the datasource metadata
- Bring up the new interface (or apply global network configuration)
- Save the updated metadata back to the pickle cache

Also scattered in some unrelated typing to make my life easier.
```

## Additional Context
Currently enabled for openstack and EC2. Haven't tested anywhere else. Testing via LXD won't be easy/possible. New ethernet devices on LXD show up as one device and is then quickly moved to another device, so the udev add event doesn't give a valid device name. That may be a scenario we want to handle at some point, but doesn't seem necessary right now.

## Test Steps
Created the following cloud-config:
```
#cloud-config
updates:
  network:
    when: ['hotplug']
```
then:
```
packages/bddeb
openstack server create --flavor m1.medium --image  cda80fdb-aa6a-4f82-8c89-352d6849766e --key-name <my_key> --nic net-id=<my_network> <server_name> --wait --user-data /my-user-data.yaml
openstack server add floating ip <id> 66650537-fbbb-4932-82fb-1b0ed9361cb2
scp cloud-init_all.deb ubuntu@<ip>:/tmp/cloud-init.deb
```
on instance:
```
dpkg -i /tmp/cloud-init.deb
cloud-init clean --logs --reboot
```
once the instance is back up, on the host:
```
openstack port create --network <network> <port_name>
openstack server add port <server_name> <port_name>
```
`ip a` should show the new nic configured appropriately with the relevant logs at the bottom of cloud-init.log


## Checklist:
